IOResources#toRetainableByteBuffer data loss when using resources without path

**Jetty version(s)**
12.1.9

**Jetty Environment**
core

**HTTP version**
not-relevant

**Java version/vendor** `(use: java -version)`
17

**OS type/version**
any

**Description**
I'm working on https://github.com/ops4j/org.ops4j.pax.web project which implements OSGi CMPN web-related specifications consistently for Jetty, Tomcat and Undertow servers.

We had a weird issues visible at this stack trace:
```
 - OSGi - 13:47:40.526 [qtp1791035325-54] WARN  (Response.java:652) org.eclipse.jetty.server.Response - writeError: status=500, message=java.io.IOException: written 872337 < 872817 content-length, response=oejsi.HttpChannelState$ErrorResponse@6d8fd6b7{500,GET@6a85c2db http://127.0.0.1:8181/documentation/ HTTP/1.1}
java.io.IOException: written 872337 < 872817 content-length
	at org.eclipse.jetty.server.internal.HttpChannelState$ChannelResponse.write(HttpChannelState.java:1385)
	at org.eclipse.jetty.server.Response$Wrapper.write(Response.java:841)
	at org.eclipse.jetty.server.handler.ContextResponse.write(ContextResponse.java:56)
	at org.eclipse.jetty.ee8.nested.HttpChannel.send(HttpChannel.java:851)
	at org.eclipse.jetty.ee8.nested.HttpChannel.sendResponse(HttpChannel.java:831)
	at org.eclipse.jetty.ee8.nested.HttpChannel.write(HttpChannel.java:905)
	at org.eclipse.jetty.ee8.nested.HttpOutput.channelWrite(HttpOutput.java:298)
	at org.eclipse.jetty.ee8.nested.HttpOutput.lambda$sendContent$1(HttpOutput.java:1140)
	at org.eclipse.jetty.http.content.CachingHttpContentFactory$CachedHttpContent.writeTo(CachingHttpContentFactory.java:391)
	at org.eclipse.jetty.ee8.nested.HttpOutput.sendContent(HttpOutput.java:1142)
	at org.eclipse.jetty.ee8.nested.HttpOutput.sendContent(HttpOutput.java:1033)
	at org.eclipse.jetty.ee8.nested.ResourceService.sendData(ResourceService.java:612)
	at org.eclipse.jetty.ee8.nested.ResourceService.doGet(ResourceService.java:281)
	at org.ops4j.pax.web.service.jetty.internal.web.DefaultServlet.doGet(DefaultServlet.java:478)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:497)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:584)
...
```

Pax Web configures Jetty Default/Resource servlet to serve static resources, which may come from OSGi bundles (so Jetty URL Resources with `bundle:` scheme are used).

I've finally reproduced the issue without any OSGi code and narrowed down the problem to `org.eclipse.jetty.io.RetainableByteBuffer.DynamicCapacity#append()` method - it doesn't mark the passed `org.eclipse.jetty.io.RetainableByteBuffer` as retained, so the buffer is reused in the loop inside `org.eclipse.jetty.io.IOResources#toRetainableByteBuffer()`.
The problem is here:
https://github.com/jetty/jetty.project/blob/a2267a1b9fe8c926e7f480956a68d0f86a0c273a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java#L118

`java.nio.ByteBuffer#array()` is passed to read operation, however the aggregate buffer may already have non-zero `java.nio.Buffer#position()`, so even if some data is read, `java.nio.Buffer#remaining()` returns zero (as the position may exceed the limit).


**How to reproduce?**
See the updated `org.eclipse.jetty.io.IOResourcesTest#testToRetainableByteBuffer()` in this PR:
* #15032

Thanks for feedback.
I've rebased my PR and commented out the line in `IOResources.java` which sets position to 0. and just run the test and got:

```
java.lang.AssertionError: 
Expected: is <8320>
     but: was <8317>
Expected :is <8320>
Actual   :<8317>
<Click to see difference>


	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
	at org.eclipse.jetty.io.IOResourcesTest.testToRetainableByteBuffer(IOResourcesTest.java:197)
...
```

When debugging the while loop in `org.eclipse.jetty.io.IOResources#toRetainableByteBuffer()`, the input stream is an instance of `sun.net.www.protocol.jar.JarURLConnection$JarURLInputStream` and the test calls its `read()` method 3 times, getting these sizes:
* 8192 (full buffer). the buffer has _retained_ flag, so its released and new one will be acquired in next loop)
* 125 bytes. I'm afraid this may be implementation detail, though I tried to ensured the ZIP file is created in reproducible way.
  * in this case, before appending the read data, and before `buffer.getByteBuffer().limit(125)` is called, the buffer state is `{java.nio.HeapByteBuffer@2940} "java.nio.HeapByteBuffer[pos=0 lim=0 cap=8192]"`
  * inside `org.eclipse.jetty.io.RetainableByteBuffer.DynamicCapacity#append()`:
    * `org.eclipse.jetty.io.RetainableByteBuffer#remaining()` returns 125 (because of the above `limit()` call)
    * current `org.eclipse.jetty.io.RetainableByteBuffer.DynamicCapacity#_aggregate` is null
    * `org.eclipse.jetty.io.RetainableByteBuffer.DynamicCapacity#shouldAggregate()` returns true, because `size < buffer.capacity() / 64` is true
    * so `org.eclipse.jetty.io.RetainableByteBuffer.DynamicCapacity#append(java.nio.ByteBuffer)` is called with the heap buffer of this state: `{java.nio.HeapByteBuffer@2940} "java.nio.HeapByteBuffer[pos=0 lim=125 cap=8192]"`
    * `org.eclipse.jetty.io.RetainableByteBuffer.DynamicCapacity#_aggregate` is acquired, because it was null, with `aggregateSize = Math.max(length = 125, aggregateSize = 125)` - mind the 3 bytes difference
    * the _aggregate is added to _buffers
    * `_aggregate.append(java.nio.ByteBuffer)` is called with bytes being the heap buffer with pos=0, lim=125, cap=8192
  * this time the buffere is NOT _retained_, so its reused in the next pass
* 3 bytes are finally read from the zip stream into buffer.getByteBuffer().array()
  * the buffer is at state `{java.nio.HeapByteBuffer@2940} "java.nio.HeapByteBuffer[pos=125 lim=125 cap=8192]"`
  * calling now `buffer.getByteBuffer().limit(3)` changes the state to: pos=3, lim=3, cap=8192 effectively making this buffer useless.

I use ArrayByteBufferPool.Tracking (as in the test) and there's a long suppressed exception printed in the test:

```
Suppressed: java.lang.AssertionError: Leaks: oeji.ArrayByteBufferPool$Tracking$TrackedBuffer@db57326 of 8192 bytes on 2026-05-14T11:06:32.986536356Z wrapping 1
 jn.HeapByteBuffer@12c7a01b[p=8192,l=8192,c=8192,r=0]={d4\xB6\xD3d\xA3\xCcm\x92...:\x0e\xC2/a=o<<<>>>}
 acquired at java.lang.Throwable: Acquired by main
	at org.eclipse.jetty.io.ArrayByteBufferPool$Tracking$TrackedBuffer.<init>(ArrayByteBufferPool.java:1068)
	at org.eclipse.jetty.io.ArrayByteBufferPool$Tracking.acquire(ArrayByteBufferPool.java:1020)
	at org.eclipse.jetty.io.ByteBufferPool$Sized.acquire(ByteBufferPool.java:162)
	at org.eclipse.jetty.io.IOResources.toRetainableByteBuffer(IOResources.java:117)
	at org.eclipse.jetty.io.IOResourcesTest.testToRetainableByteBuffer(IOResourcesTest.java:196)
...
```

Not sure how the unpredictability of reading from the zip input stream affects the test...

here's my Java version:
```
openjdk version "17.0.19" 2026-04-21
OpenJDK Runtime Environment Temurin-17.0.19+10 (build 17.0.19+10)
OpenJDK 64-Bit Server VM Temurin-17.0.19+10 (build 17.0.19+10, mixed mode, sharing)
```

I checked your change and it works perfectly!
I also verified my initial place where it was failing - changed from:
```
[INFO] Running org.ops4j.pax.web.itest.jetty.war.KarafDocWarIntegrationTest
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.714 s <<< FAILURE! -- in org.ops4j.pax.web.itest.jetty.war.KarafDocWarIntegrationTest
[ERROR] org.ops4j.pax.web.itest.jetty.war.KarafDocWarIntegrationTest.testSlash -- Time elapsed: 0.609 s <<< FAILURE!
java.lang.AssertionError: 
Result is not conforming to expected definitions!
Unexpected HttpStatusCode returned! Expected '200', but was '500'
	Response-Content mismatch: Response must contain 'Apache Karaf', but is <html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"/>
<title>Error 500 java.io.IOException: written 872337 &lt; 872817 content-length</title>
</head>
<body>
<h2>HTTP ERROR 500 java.io.IOException: written 872337 &lt; 872817 content-length</h2>
<table>
<tr><th>URI:</th><td>http://127.0.0.1:8181/documentation/</td></tr>
<tr><th>STATUS:</th><td>500</td></tr>
<tr><th>MESSAGE:</th><td>java.io.IOException: written 872337 &lt; 872817 content-length</td></tr>
</table>

</body>
</html>
```

to:
```
[INFO] Running org.ops4j.pax.web.itest.jetty.war.KarafDocWarIntegrationTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.601 s -- in org.ops4j.pax.web.itest.jetty.war.KarafDocWarIntegrationTest
```

Thanks again!

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java
@@ -122,8 +122,16 @@ public class IOResources
                 retainableByteBuffer.append(buffer);
                 if (buffer.isRetained())
                 {
+                    // buffer has been retained by DynamicCapacity, fetch
+                    // a new one on the next loop iteration.
                     buffer.release();
                     buffer = null;
+                }
+                else
+                {
+                    // buffer has been copied by DynamicCapacity, clear it
+                    // before reusing it for the next loop iteration.
+                    buffer.clear();
                 }
             }
             return retainableByteBuffer;

--- a/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/IOResourcesTest.java
+++ b/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/IOResourcesTest.java
@@ -13,6 +13,7 @@
 
 package org.eclipse.jetty.io;
 
+import java.io.ByteArrayInputStream;
 import java.io.InputStream;
 import java.net.URI;
 import java.nio.ByteBuffer;
@@ -34,13 +35,16 @@ import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
 import org.junit.jupiter.params.provider.MethodSource;
 
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.empty;
 import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
 import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.params.provider.Arguments.arguments;
 
 public class IOResourcesTest
 {
@@ -141,32 +145,126 @@ public class IOResourcesTest
         }
     }
 
-    public static Stream<Resource> all() throws Exception
+    // This Resource impl has getPath() return null so newInputStream() is the only way to read its contents,
+    // plus the returned input stream can only read 1 byte at a time.
+    private static class TestInputStreamResource extends Resource
+    {
+        private final URI uri;
+        private final byte[] buffer;
+
+        public TestInputStreamResource(URI uri, byte[] buffer)
+        {
+            this.uri = uri;
+            this.buffer = buffer;
+        }
+
+        @Override
+        public boolean exists()
+        {
+            return true;
+        }
+
+        @Override
+        public long length()
+        {
+            return buffer.length;
+        }
+
+        @Override
+        public Path getPath()
+        {
+            return null;
+        }
+
+        @Override
+        public InputStream newInputStream()
+        {
+            return new ByteArrayInputStream(buffer)
+            {
+                @Override
+                public synchronized int read(byte[] b, int off, int len)
+                {
+                    int read = read();
+                    if (read == -1)
+                        return -1;
+                    b[off] = (byte)read;
+                    return 1;
+                }
+            };
+        }
+
+        @Override
+        public boolean isDirectory()
+        {
+            return false;
+        }
+
+        @Override
+        public boolean isReadable()
+        {
+            return true;
+        }
+
+        @Override
+        public URI getURI()
+        {
+            return uri;
+        }
+
+        @Override
+        public String getName()
+        {
+            return uri.getPath();
+        }
+
+        @Override
+        public String getFileName()
+        {
+            return uri.getPath();
+        }
+
+        @Override
+        public Resource resolve(String subUriPath)
+        {
+            return null;
+        }
+    }
+
+    public static Stream<Arguments> resourcesWithBytes() throws Exception
     {
         Path testResourcePath = MavenTestingUtils.getTestResourcePath("keystore.p12");
-
-        URI resourceUri = testResourcePath.toUri();
+        URI testResourceUri = testResourcePath.toUri();
+        byte[] testResourceBytes = Files.readAllBytes(testResourcePath);
         return Stream.of(
-            ResourceFactory.root().newResource(resourceUri),
-            ResourceFactory.root().newMemoryResource(resourceUri.toURL()),
-            ResourceFactory.root().newResource(MavenTestingUtils.getTestResourcePath("zero")),
-            ResourceFactory.root().newResource(MavenTestingUtils.getTestResourcePath("one")),
-            new URLResourceFactory().newResource(resourceUri),
-            new TestContentSourceFactoryResource(resourceUri, Files.readAllBytes(testResourcePath))
+            arguments(ResourceFactory.root().newResource(testResourceUri), testResourceBytes),
+            arguments(ResourceFactory.root().newResource(testResourceUri), testResourceBytes),
+            arguments(ResourceFactory.root().newMemoryResource(testResourceUri.toURL()), testResourceBytes),
+            arguments(ResourceFactory.root().newResource(MavenTestingUtils.getTestResourcePath("zero")), BufferUtil.EMPTY_BYTES),
+            arguments(ResourceFactory.root().newResource(MavenTestingUtils.getTestResourcePath("one")), Files.readAllBytes(MavenTestingUtils.getTestResourcePath("one"))),
+            arguments(new URLResourceFactory().newResource(testResourceUri), testResourceBytes),
+            arguments(new TestInputStreamResource(testResourceUri, testResourceBytes), testResourceBytes),
+            arguments(new TestContentSourceFactoryResource(testResourceUri, testResourceBytes), testResourceBytes)
         );
     }
 
+    public static Stream<Resource> resources() throws Exception
+    {
+        return resourcesWithBytes().map(arguments -> (Resource)arguments.get()[0]);
+    }
+
     @ParameterizedTest
-    @MethodSource("all")
-    public void testToRetainableByteBuffer(Resource resource)
+    @MethodSource("resourcesWithBytes")
+    public void testToRetainableByteBuffer(Resource resource, byte[] resourceBytes)
     {
         RetainableByteBuffer retainableByteBuffer = IOResources.toRetainableByteBuffer(resource, bufferPool);
         assertThat(retainableByteBuffer.remaining(), is((int)resource.length()));
+        assertThat(retainableByteBuffer.size(), is(resource.length()));
+        assertArrayEquals(resourceBytes, BufferUtil.toArray(retainableByteBuffer.getByteBuffer()));
         retainableByteBuffer.release();
     }
 
     @ParameterizedTest
-    @MethodSource("all")
+    @MethodSource("resources")
     public void testAsContentSource(Resource resource) throws Exception
     {
         TestSink sink = new TestSink();
@@ -181,7 +279,7 @@ public class IOResourcesTest
     }
 
     @ParameterizedTest
-    @MethodSource("all")
+    @MethodSource("resources")
     public void testAsContentSourceWithOffset(Resource resource) throws Exception
     {
         TestSink sink = new TestSink();
@@ -202,7 +300,7 @@ public class IOResourcesTest
     }
 
     @ParameterizedTest
-    @MethodSource("all")
+    @MethodSource("resources")
     public void testAsContentSourceWithLength(Resource resource) throws Exception
     {
         TestSink sink = new TestSink();
@@ -217,7 +315,7 @@ public class IOResourcesTest
     }
 
     @ParameterizedTest
-    @MethodSource("all")
+    @MethodSource("resources")
     public void testAsContentSourceWithOffsetAndLength(Resource resource) throws Exception
     {
         TestSink sink = new TestSink();
@@ -235,7 +333,7 @@ public class IOResourcesTest
     }
 
     @ParameterizedTest
-    @MethodSource("all")
+    @MethodSource("resources")
     public void testCopy(Resource resource) throws Exception
     {
         TestSink sink = new TestSink();
@@ -249,7 +347,7 @@ public class IOResourcesTest
     }
 
     @ParameterizedTest
-    @MethodSource("all")
+    @MethodSource("resources")
     public void testCopyWithOffset(Resource resource) throws Exception
     {
         TestSink sink = new TestSink();
@@ -264,7 +362,7 @@ public class IOResourcesTest
     }
 
     @ParameterizedTest
-    @MethodSource("all")
+    @MethodSource("resources")
     public void testCopyWithLength(Resource resource) throws Exception
     {
         TestSink sink = new TestSink();
@@ -279,7 +377,7 @@ public class IOResourcesTest
     }
 
     @ParameterizedTest
-    @MethodSource("all")
+    @MethodSource("resources")
     public void testCopyWithOffsetAndLength(Resource resource) throws Exception
     {
         TestSink sink = new TestSink();
@@ -295,7 +393,7 @@ public class IOResourcesTest
     }
 
     @ParameterizedTest
-    @MethodSource("all")
+    @MethodSource("resources")
     public void testOutOfRangeOffset(Resource resource)
     {
         TestSink sink = new TestSink();
@@ -305,7 +403,7 @@ public class IOResourcesTest
     }
 
     @ParameterizedTest
-    @MethodSource("all")
+    @MethodSource("resources")
     public void testOutOfRangeOffsetWithZeroLength(Resource resource) throws Exception
     {
         TestSink sink = new TestSink();

